tests/utils/git_test.py

```diff
@@ -17,63 +17,139 @@
 
 import logging
 import os
+import re
+import shutil
 from unittest import mock
 
 import pytest
 
 from esrally import exceptions
-from esrally.utils import git
+from esrally.utils import git, io, process
 
 
+# pylint: disable=too-many-public-methods
 class TestGit:
+    @classmethod
+    def setup_class(cls):
+        cls.remote_repo = "esrally"
+        cls.local_branch = "rally-unit-test-local-only-branch"
+        cls.remote_branch = "rally-unit-test-remote-only-branch"
+        cls.rebase_branch = "rally-unit-test-rebase-branch"
+
+        # this is assuming that nobody stripped the git repo info in their Rally working copy
+        cls.original_src_dir = os.path.dirname(os.path.dirname(os.path.dirname(__file__)))
+        cls.local_tmp_src_dir = io.escape_path("/tmp/rally-unit-test-local-dir")
+        cls.remote_tmp_src_dir = io.escape_path("/tmp/rally-unit-test-remote-dir")
+        cls.tmp_clone_dir = io.escape_path("/tmp/rally-unit-test-clone-dir")
+
+        # delete any pre-existing dirs (maybe last test setup cancelled early)
+        shutil.rmtree(TestGit.local_tmp_src_dir, ignore_errors=True)
+        shutil.rmtree(TestGit.remote_tmp_src_dir, ignore_errors=True)
+        shutil.rmtree(TestGit.tmp_clone_dir, ignore_errors=True)
+
+        # create tmp git repos
+        shutil.copytree(TestGit.original_src_dir, TestGit.local_tmp_src_dir)
+        # stash any current changes in copied dir
+        process.run_subprocess_with_logging(f"git -C {TestGit.local_tmp_src_dir} stash")
+        shutil.copytree(TestGit.original_src_dir, TestGit.remote_tmp_src_dir)
+        # so we can restore the working tree after some tests
+        cls.starting_branch = git.current_branch(TestGit.local_tmp_src_dir)
+
+        # setup test branches
+        process.run_subprocess_with_logging(f"git -C {TestGit.local_tmp_src_dir} branch {TestGit.local_branch}")
+        process.run_subprocess_with_logging(f"git -C {TestGit.remote_tmp_src_dir} branch {TestGit.remote_branch}")
+        cls.remote_branch_hash = process.run_subprocess_with_output(
+            f"git -C {TestGit.remote_tmp_src_dir} rev-parse {TestGit.remote_branch}"
+        )[0]
+        # setup remote
+        process.run_subprocess_with_logging(
+            f"git -C {TestGit.local_tmp_src_dir} remote add {TestGit.remote_repo} {os.path.join(TestGit.remote_tmp_src_dir, '.git')}"
+        )
+
+        # fetch branches from remote
+        git.fetch(TestGit.local_tmp_src_dir, remote=TestGit.remote_repo)
+
+    @classmethod
+    def teardown_class(cls):
+        # delete copy of git dir
+        shutil.rmtree(TestGit.local_tmp_src_dir)
+        shutil.rmtree(TestGit.remote_tmp_src_dir)
+
+    @pytest.fixture
+    def teardown_clone(self):
+        yield
+        shutil.rmtree(TestGit.tmp_clone_dir)
```

Comment:
This seems quite implicit, why is this a fixture?

Reply:
Addressed in https://github.com/elastic/rally/pull/1602/commits/529ea0a8dd8a9cdd47b33890eadbd05ea5cf13e4 and https://github.com/elastic/rally/pull/1602/commits/8f50ce76806a43d42ecf30fc300e7ff48cf724b8